src/main/java/com/amazon/ion/impl/bin/TimestampEncoder_1_1.kt

```diff
@@ -0,0 +1,233 @@
+        // timestamp fields and then have one branch at the end to write the correct number of bits.
+        // This also keeps the code a little shorter.
+
+        // TODO(perf) revisit this to see if this is slower than branching earlier.
+
+        var bits = (value.year - 1970L)
+        bits = bits or ((value.month.toLong()) shl Ion_1_1_Constants.S_TIMESTAMP_MONTH_BIT_OFFSET)
+        bits = bits or ((value.day.toLong()) shl Ion_1_1_Constants.S_TIMESTAMP_DAY_BIT_OFFSET)
+        bits = bits or ((value.hour.toLong()) shl Ion_1_1_Constants.S_TIMESTAMP_HOUR_BIT_OFFSET)
+        bits = bits or ((value.minute.toLong()) shl Ion_1_1_Constants.S_TIMESTAMP_MINUTE_BIT_OFFSET)
+        val localOffset = (value.localOffset.toLong() / 15) + (14 * 4)
+        bits = bits or ((localOffset and Ion_1_1_Constants.LEAST_SIGNIFICANT_7_BITS) shl Ion_1_1_Constants.S_O_TIMESTAMP_OFFSET_BIT_OFFSET)
+        bits = bits or ((value.second.toLong()) shl Ion_1_1_Constants.S_O_TIMESTAMP_SECOND_BIT_OFFSET)
+
+        buffer.writeFixedIntOrUInt(bits, 5)
+
+        // The fractional seconds bits will be put into a separate long because we need nine bytes total
+        // if there are nanoseconds (which is too much for one long) and the boundary between the seconds
+        // and fractional seconds subfields conveniently aligns with a byte boundary.
+        var fractionBits = 0L
+
+        val size = when (implicitOpcode) {
+            OpCode.TIMESTAMP_MINUTE_PRECISION_WITH_OFFSET,
+            OpCode.TIMESTAMP_SECOND_PRECISION_WITH_OFFSET -> {
+                5
+            }
+            OpCode.TIMESTAMP_MILLIS_PRECISION_WITH_OFFSET -> {
+                fractionBits = value.zFractionalSecond.unscaledValue().toLong()
+                buffer.writeFixedIntOrUInt(fractionBits, 2)
+                7
+            }
+            OpCode.TIMESTAMP_MICROS_PRECISION_WITH_OFFSET -> {
+                fractionBits = value.zFractionalSecond.unscaledValue().toLong()
+                buffer.writeFixedIntOrUInt(fractionBits, 3)
+                8
+            }
+            OpCode.TIMESTAMP_NANOS_PRECISION_WITH_OFFSET -> {
+                fractionBits = value.zFractionalSecond.unscaledValue().toLong()
+                buffer.writeFixedIntOrUInt(fractionBits, 4)
+                9
+            }
+            else -> throw IllegalStateException("This is unreachable!")
+        }
+        return size
+    }
+
+    /**
+     * Writes a long-form timestamp.
+     * Value may not be null.
+     * Only visible for testing. If calling from outside this class, use writeTimestampValue instead.
+     */
+    @JvmStatic
+    internal fun writeLongFormTimestampBody(buffer: WriteBuffer, value: Timestamp): Int {
+        var bits = value.year.toLong()
+        if (value.precision == Timestamp.Precision.YEAR) {
+            buffer.writeFlexUInt(2)
+            buffer.writeFixedIntOrUInt(bits, 2)
```

Comment:
This pattern recurs frequently. I wonder if it would be (marginally) faster to have a method variant that does both in one step.